hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h

```diff
@@ -90,54 +90,45 @@ namespace hazelcast {
                                     const std::shared_ptr<spi::impl::ClientInvocation> invocation) override {
                         check_connection(connection, invocation);
                         auto message = invocation->get_client_message();
-                        socket_strand_.post([=]() {
+                        socket_strand_.post([connection, invocation, message, this]() {
                             if (!check_connection(connection, invocation)) {
                                 return;
                             }
 
                             bool success;
                             int64_t message_call_id;
-                            do {
-                                auto call_id = ++call_id_counter_;
-                                struct correlation_id {
-                                    int32_t connnection_id;
-                                    int32_t call_id;
-                                };
-                                union {
-                                    int64_t id;
-                                    correlation_id composed_id;
-                                } c_id_union;
-                                c_id_union.composed_id = {connection->get_connection_id(), call_id};
-                                message_call_id = c_id_union.id;
-                                message->set_correlation_id(c_id_union.id);
-                                success = connection->invocations.insert({message_call_id, invocation}).second;
-                            } while (!success);
-
-                            auto handler = [=](const boost::system::error_code &ec,
-                                               std::size_t bytes_written) {
-                                if (ec) {
-                                    auto message = (boost::format{
-                                            "Error %1% during invocation write for %2% on connection %3%"} %
-                                                    ec % *invocation % *connection).str();
-                                    connection->close(message);
-                                } else {
-                                    // update the connection write time
-                                    connection->last_write_time(std::chrono::steady_clock::now());
-                                }
+                            auto call_id = ++call_id_counter_;
+                            struct correlation_id {
+                                int32_t connnection_id;
+                                int32_t call_id;
                             };
+                            union {
+                                int64_t id;
+                                correlation_id composed_id;
+                            } c_id_union;
+                            c_id_union.composed_id = {connection->get_connection_id(), call_id};
+                            message_call_id = c_id_union.id;
```

Comment:
It can be done. Endianness may not be a problem since the correlation id is only unique per connection and we always use the incremented `call_id_counter_`. Does that look simpler? Is that one more efficient or current impl? I am not sure, let me try to evaluate. I can make it a method call, yes.

Reply:
I've checked both options on the compiler explorer: https://godbolt.org/z/5ffdjcob8

I am definitely not an expert in C++ or assembly, but I believe the shift + OR could be very slightly more efficient (my reasoning is, there are less stack MOV instructions with the former) but I don't think that's important at all.

Reply:
Just a note using `-O1` with GCC optimizes both approaches to the same code: https://godbolt.org/z/ncYaeMMqE

Reply:
Surprisingly MSVC optimizes the struct+union one a lot better: https://godbolt.org/z/nT3Kre9fh